File: tools/profile_trace_line_reduce.cpp
#include <algorithm>
#include <cctype>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

struct Stat {
  long long count = 0;
  long double total_us = 0.0;
  double max_us = 0.0;
};

static std::string extract_quoted_after(const std::string& line, const std::string& key) {
  const auto key_pos = line.find(key);
  if (key_pos == std::string::npos) return "";
  const auto first = line.find('"', key_pos + key.size());
  if (first == std::string::npos) return "";
  const auto second = line.find('"', first + 1);
  if (second == std::string::npos) return "";
  return line.substr(first + 1, second - first - 1);
}

static bool extract_number_after(
    const std::string& line, const std::string& key, double* value) {
  const auto key_pos = line.find(key);
  if (key_pos == std::string::npos) return false;
  auto pos = key_pos + key.size();
  while (pos < line.size() && (line[pos] == ' ' || line[pos] == ':')) ++pos;
  char* end = nullptr;
  const double parsed = std::strtod(line.c_str() + pos, &end);
  if (end == line.c_str() + pos) return false;
  *value = parsed;
  return true;
}

static std::string lower_copy(std::string s) {
  std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) {
    return static_cast<char>(std::tolower(c));
  });
  return s;
}

static bool contains_any(const std::string& hay, const std::vector<std::string>& needles) {
  for (const auto& needle : needles) {
    if (hay.find(needle) != std::string::npos) return true;
  }
  return false;
}

static bool starts_with(const std::string& value, const std::string& prefix) {
  return value.rfind(prefix, 0) == 0;
}

static std::vector<std::string> groups_for(const std::string& name, const std::string& cat) {
  const std::string hay = lower_copy(name + " " + cat);
  std::vector<std::string> groups;
  if (starts_with(name, "ProfilerStep")) groups.push_back("profiler_step");
  if (starts_with(name, "dsa.") || starts_with(name, "hisa.") || starts_with(name, "moe.") ||
      starts_with(name, "fsdp.") || starts_with(name, "pipeline.")) {
    groups.push_back("custom_range");
  }
  if (hay.find("hisa") != std::string::npos || hay.find("_hisa") != std::string::npos) {
    groups.push_back("hisa");
  }
  if (hay.find("dsa") != std::string::npos || hay.find("dsattention") != std::string::npos) {
    groups.push_back("dsa");
  }
  if (contains_any(hay, {"moe", "expert", "deepep"})) groups.push_back("moe");
  if (contains_any(hay, {"fsdp", "all_gather_params"})) groups.push_back("fsdp");
  if (contains_any(
          hay,
          {"nccl", "allreduce", "all_reduce", "alltoall", "all_to_all", "reduce_scatter",
           "broadcast"})) {
    groups.push_back("comm");
  }
  if (contains_any(hay, {"memcpy", "memset", "copy"}) || name == "aten::copy_") {
    groups.push_back("copy_mem");
  }
  if (starts_with(name, "aten::")) groups.push_back("aten");
  if (hay.find("kernel") != std::string::npos && !starts_with(name, "cudaLaunchKernel")) {
    groups.push_back("cuda_kernel");
  }
  if (hay.find("cuda") != std::string::npos || starts_with(name, "cuda")) {
    groups.push_back("cuda_runtime");
  }
  if (groups.empty()) groups.push_back("other");
  return groups;
}

static void add_stat(std::unordered_map<std::string, Stat>& stats, const std::string& key, double dur) {
  auto& stat = stats[key];
  stat.count += 1;
  stat.total_us += dur;
  stat.max_us = std::max(stat.max_us, dur);
}

static std::string json_escape(const std::string& input) {
  std::string out;
  out.reserve(input.size() + 8);
  for (const char c : input) {
    switch (c) {
      case '\\': out += "\\\\"; break;
      case '"': out += "\\\""; break;
      case '\n': out += "\\n"; break;
      case '\r': out += "\\r"; break;
      case '\t': out += "\\t"; break;
      default: out += c; break;
    }
  }
  return out;
}

static void write_stats_object(const std::unordered_map<std::string, Stat>& stats) {
  std::vector<std::string> keys;
  keys.reserve(stats.size());
  for (const auto& kv : stats) keys.push_back(kv.first);
  std::sort(keys.begin(), keys.end());
  std::cout << "{";
  bool first = true;
  for (const auto& key : keys) {
    if (!first) std::cout << ",";
    first = false;
    const auto& stat = stats.at(key);
    std::cout << "\n      \"" << json_escape(key) << "\": {"
              << "\"count\": " << stat.count
              << ", \"total_us\": " << std::fixed << std::setprecision(3)
              << static_cast<double>(stat.total_us)
              << ", \"max_us\": " << stat.max_us << "}";
  }
  if (!keys.empty()) std::cout << "\n    ";
  std::cout << "}";
}

int main(int argc, char** argv) {
  if (argc != 2) {
    std::cerr << "usage: profile_trace_line_reduce_cpp <rank>\n";
    return 2;
  }
  const int rank = std::atoi(argv[1]);
  std::unordered_map<std::string, Stat> groups;
  std::unordered_map<std::string, std::unordered_map<std::string, Stat>> names;
  std::string current_name;
  std::string current_cat;
  std::string line;
  long long events = 0;
  double first_ts = 0.0;
  double last_ts = 0.0;
  bool have_ts = false;

  while (std::getline(std::cin, line)) {
    if (line.find("\"name\":") != std::string::npos) {
      const std::string name = extract_quoted_after(line, "\"name\"");
      if (!name.empty()) {
        current_name = name;
        current_cat = extract_quoted_after(line, "\"cat\"");
      }
    }
    if (current_name.empty() || line.find("\"dur\":") == std::string::npos) continue;
    double dur = 0.0;
    if (!extract_number_after(line, "\"dur\"", &dur)) continue;
    double ts = 0.0;
    if (extract_number_after(line, "\"ts\"", &ts)) {
      if (!have_ts) {
        first_ts = ts;
        last_ts = ts;
        have_ts = true;
      } else {
        first_ts = std::min(first_ts, ts);
        last_ts = std::max(last_ts, ts);
      }
    }
    events += 1;
    for (const auto& group : groups_for(current_name, current_cat)) {
      add_stat(groups, group, dur);
      add_stat(names[group], current_name, dur);
    }
    current_name.clear();
    current_cat.clear();
  }

  std::cout << "{\n  \"rank\": " << rank << ",\n";
  std::cout << "  \"events\": " << events << ",\n";
  if (have_ts) {
    std::cout << "  \"first_ts\": " << std::fixed << std::setprecision(3) << first_ts << ",\n";
    std::cout << "  \"last_ts\": " << std::fixed << std::setprecision(3) << last_ts << ",\n";
  } else {
    std::cout << "  \"first_ts\": null,\n  \"last_ts\": null,\n";
  }
  std::cout << "  \"groups\": ";
  write_stats_object(groups);
  std::cout << ",\n  \"names\": {";
  std::vector<std::string> group_keys;
  group_keys.reserve(names.size());
  for (const auto& kv : names) group_keys.push_back(kv.first);
  std::sort(group_keys.begin(), group_keys.end());
  bool first_group = true;
  for (const auto& group : group_keys) {
    if (!first_group) std::cout << ",";
    first_group = false;
    std::cout << "\n    \"" << json_escape(group) << "\": ";
    write_stats_object(names.at(group));
  }
  if (!group_keys.empty()) std::cout << "\n  ";
  std::cout << "}\n}\n";
  return 0;
}

File: tools/profile_trace_line_reduce.py
#!/usr/bin/env python3
# Copyright (c) 2026, NVIDIA CORPORATION. All rights reserved.

"""Fast line-oriented reducer for pretty-printed PyTorch Chrome traces.

PyTorch's exported traces in this run are multi-line JSON, but each event's
``cat``/``name`` line is followed by a ``ts``/``dur`` line. Full JSON parsing is
too slow for 100GB+ decompressed traces; this reducer extracts the fields we
need directly while streaming stdin.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from collections import defaultdict
from typing import BinaryIO


CAT_RE = re.compile(rb'"cat":\s*"([^"]*)"')
NAME_RE = re.compile(rb'"name":\s*"([^"]*)"')
DUR_RE = re.compile(rb'"dur":\s*([-+0-9.eE]+)')
TS_RE = re.compile(rb'"ts":\s*([-+0-9.eE]+)')


def empty_stat() -> dict[str, float | int]:
    return {"count": 0, "total_us": 0.0, "max_us": 0.0}


def add_stat(stats: dict[str, dict[str, float | int]], key: str, dur: float) -> None:
    stat = stats[key]
    stat["count"] = int(stat["count"]) + 1
    stat["total_us"] = float(stat["total_us"]) + dur
    stat["max_us"] = max(float(stat["max_us"]), dur)


def decode(raw: bytes) -> str:
    return raw.decode("utf-8", errors="replace")


def groups_for(name: str, cat: str) -> list[str]:
    hay = f"{name} {cat}".lower()
    out: list[str] = []
    if name.startswith("ProfilerStep"):
        out.append("profiler_step")
    if name.startswith(("dsa.", "hisa.", "moe.", "fsdp.", "pipeline.")):
        out.append("custom_range")
    if "hisa" in hay or "_hisa" in hay:
        out.append("hisa")
    if "dsa" in hay or "dsattention" in hay:
        out.append("dsa")
    if "moe" in hay or "expert" in hay or "deepep" in hay:
        out.append("moe")
    if "fsdp" in hay or "all_gather_params" in hay:
        out.append("fsdp")
    if any(token in hay for token in ("nccl", "allreduce", "all_reduce", "alltoall", "all_to_all", "reduce_scatter", "broadcast")):
        out.append("comm")
    if any(token in hay for token in ("memcpy", "memset", "copy")) or name == "aten::copy_":
        out.append("copy_mem")
    if name.startswith("aten::"):
        out.append("aten")
    if "kernel" in hay and not name.startswith("cudaLaunchKernel"):
        out.append("cuda_kernel")
    if "cuda" in hay or name.startswith("cuda"):
        out.append("cuda_runtime")
    if not out:
        out.append("other")
    return out


def reduce_stream(stream: BinaryIO, rank: int) -> dict[str, object]:
    groups: dict[str, dict[str, float | int]] = defaultdict(empty_stat)
    names: dict[str, dict[str, dict[str, float | int]]] = defaultdict(lambda: defaultdict(empty_stat))
    events = 0
    first_ts: float | None = None
    last_ts: float | None = None
    current_name: str | None = None
    current_cat = ""

    for line in stream:
        if b'"name":' in line:
            name_match = NAME_RE.search(line)
            if name_match:
                current_name = decode(name_match.group(1))
                cat_match = CAT_RE.search(line)
                current_cat = decode(cat_match.group(1)) if cat_match else ""
        if current_name is None or b'"dur":' not in line:
            continue
        dur_match = DUR_RE.search(line)
        if not dur_match:
            continue
        try:
            dur = float(dur_match.group(1))
        except ValueError:
            continue
        ts_match = TS_RE.search(line)
        if ts_match:
            try:
                ts = float(ts_match.group(1))
            except ValueError:
                ts = None
            if ts is not None:
                first_ts = ts if first_ts is None else min(first_ts, ts)
                last_ts = ts if last_ts is None else max(last_ts, ts)
        events += 1
        for group in groups_for(current_name, current_cat):
            add_stat(groups, group, dur)
            add_stat(names[group], current_name, dur)
        current_name = None
        current_cat = ""

    return {
        "rank": rank,
        "events": events,
        "first_ts": first_ts,
        "last_ts": last_ts,
        "groups": groups,
        "names": names,
    }


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--rank", required=True, type=int)
    parser.add_argument("--output", required=True)
    args = parser.parse_args()

    summary = reduce_stream(sys.stdin.buffer, args.rank)
    with open(args.output, "w", encoding="utf-8") as f:
        json.dump(summary, f, indent=2, sort_keys=True)
        f.write("\n")
    print(args.output)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
